### PR TITLE
WIP: Enable 1% trace sampling by default.

### PR DESCRIPTION
**Pull request by mandarjog**
1. Enable 1% trace sampling.
2. This PR uses a deprecated method of setting these parameters until this is supported thru xDS.

### CODE REVIEW

**Comment by ldemailly** [comment]
how can I force tracing per request or set it back to 100% after this goes in ?

**Comment by mandarjog** [comment]
If trace is forced externally then Envoy will participate in that trace. This PR does not change that.
If **No** trace information is found on a request, then envoy will only initiate traces for 1% requests.

**Comment by ldemailly** [comment]
how do we plan on having some optional header to trigger tracing for troubleshooting (and in my case to calculate 0% caching worst case cost/latency)

ps: optional as it may be a security issue / dos vector

**Comment by saturnism** [comment]
is there a configuration value for the sampling rate / probability?

File: pilot/docker/random_sampling [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# per 10000 of requests that will be randomly traced.
+# See here for more information.
```
Comment (ldemailly) [inline pilot/docker/random_sampling]: here?

File: pilot/docker/Dockerfile.proxy_debug [hunk 2]
```diff
@@ -36,5 +36,8 @@ RUN useradd -m --uid 1337 istio-proxy && \
     echo "istio-proxy ALL=NOPASSWD: ALL" >> /etc/sudoers && \
     chown -R istio-proxy /var/lib/istio
 
+# Control tracing using legacy runtime API
+COPY random_sampling /etc/istio/proxy/runtime/tracing/random_sampling
```
Comment (ldemailly) [inline pilot/docker/Dockerfile.proxy_debug]: maybe that can be a configmap mount ?

Comment (mandarjog) [reply pilot/docker/Dockerfile.proxy_debug]: config map sounds good.
Will make that a separate PR.

**Comment by mandarjog** [comment]
This is potentially blocked on https://github.com/envoyproxy/envoy/issues/2710

**Comment by ldemailly** [comment]
murphy is such that any non tested feature doesn't work I suppose

**Comment by douglas-reid** [comment]
see also: https://github.com/istio/api/pull/363

**Comment by mandarjog** [comment]
All of that is true, the question is simply one of timing.
There are several PR sitting in envoy that deal with this for a some time.
We have heard from customers that a default of 100% requests traced is not good.
This PR changes that default to 1%. You can over mount a config volume and change the default.

We don't need any of the symlink swapping because we treat these as startup parameters, that are largely hidden until.

We can easily have different defaults (100-% for intro), (1% for production use thru helm charts)
This solves a problem today.

**Comment by ldemailly** [comment]
crash reproes?

**Comment by rshriram** [comment]
what would you do if the said customer wants to change the 1% to 10% without bringing down the ingress controller? This feels like a hack for one customer. While it might solve one customer's issue, it is probably going to cause issues with the rest of the folks. Not everyone is okay with 1% sampling by default. If the rest of customers want to tweak the sampling rate, we don't have a pipeline for that.

**Comment by costinm** [comment]
Please resolve conflicts - I don't think a better option will be ready for 0.8, so we should try to get this as a workaround - 100% tracing has performance impact so we need to have something.

I think it is fine to only include this in the proxyv2 - no need to add more features or risks to alpha1/v1

**Comment by geeknoid** [comment]
Do we still need this PR?

**Comment by rshriram** [comment]
Nah. Doug had another pr in istio/api but that seems to be in limbo

**Comment by douglas-reid** [comment]
I believe @jeffmendoza is still planning on moving the istio/api PR forward.

**Comment by jeffmendoza** [comment]
That is correct.

**Comment by geeknoid** [comment]
Closing...

**Pull request closed without merging**